New-feature, WiP: Ephemeral Environments for PRs — Uffizzi Integration

ShrutiC-git opened the pull request:
Thank you for contributing to Harbor!

# Comprehensive Summary of your change
This PR is adding support for preview environments for PR events (opened, sync'ed, reopened, etc) on Harbor, powered by [Uffizzi](https://www.uffizzi.com/).

This PR uses GHA workflows to build and provision preview environments. Utilizing docker-compose examples from Harbor, the PR uses docker-compose.uffizzi.yml to build open-sourced components of Harbor from source: Harbor portal, Harbor, registryctl, and jobservice. Building from source makes sure the latest changes are reflected in the preview. 
The docker-compose also has Harbor core, Harbor registry, postgres charmuseum and redis containers, utilizing images from DockerHub. 

Once a PR event is triggered — PR opened, PR synced, review requested — the build workflow will build Harbor portal, jobservice and harbor-registryctl from source along with other related services, and the preview workflow will either spin up a new preview or update an existing preview. The previews are ephemeral — so they live for only as long as needed.

A comment will be posted on the PR containing the URL of the preview environment, allowing contributors and maintainers to easily navigate to the preview.

To test this, the first step was opening a PR against my fork of Harbor, the uffizzi-build.yml file triggered the Build PR Image action. Once this action was completed, the uffizzi-preview.yml GHA-file triggered the Deploy Preview action. We use this 2-step workflow for open-source projects, to allow open-source projects to have previews for contributions from forks. [Here](https://www.uffizzi.com/preview-environments-guide/github-actions-two-stage-workflow) is a more in-depth read into why we use a two-stage workflow.

* [Here is the PR](https://github.com/ShrutiC-git/harbor/pull/8) opened against my personal fork of Habor
* [Here is the comment](https://github.com/ShrutiC-git/harbor/pull/8#issuecomment-1398012998) that was posted on my PR.
* [Here is the preview environment](https://app.uffizzi.com/github.com/ShrutiC-git/harbor/pull/8) that was created.

To log into the preview, use the following credentials:
* username: admin
* password: password

# Issue being fixed
Fixes #18006 

Please indicate you've done the following:
- [X] Well Written Title and Summary of the PR
- [X] Label the PR as needed. "release-note/ignore-for-release, release-note/new-feature, release-note/update, release-note/enhancement, release-note/community, release-note/breaking-change, release-note/docs, release-note/infra, release-note/deprecation"
- [X] Accepted the DCO. Commits without the DCO will delay acceptance.
- [X] Made sure tests are passing and test coverage is added if needed.
- [ ] Considered the docs impact and opened a new docs issue or PR with docs changes if needed in [website repository](https://github.com/goharbor/website).


cc @waveywaves

ShrutiC-git commented:
On the preview, it seems some of the APi endpoints from API Core return a 404 — namely some endpoints concerning jobservice (`api/v2.0/jobservice`). This PR is using Harbor core image from GitHub and it seems that some of the DB migration scrips for core might not be running. 

I would really appreciate some help on how the issue with some of these endpoints can be resolved!

Vad1mo commented:
Thank you for your contribution, it is excellent to have a preview environment. 
I need to investigate further, but I fear that your changes will break the current packaging and offering. 
I would suggest putting the preview config and modifications into a `preview` subdirectory, so the code for preview  does not interfere with our prod and release code.

ShrutiC-git commented:
thanks for taking a look into this, @Vad1mo! You're right — I will move the changes into its own directory so it's efficient and not intrusive. 
If we could go over the changes and other modifications over a call, that'd be really helpful design this more effectively. It'll help give more understanding into building other components which I was not able to find in the source code. Let me know if that sounds okay!

OrlinVasilev commented on config/registry/root.crt:
```diff
@@ -0,0 +1,35 @@
+-----BEGIN CERTIFICATE-----
+MIIGBzCCA++gAwIBAgIJAKB8CNqCxhr7MA0GCSqGSIb3DQEBCwUAMIGZMQswCQYD
+VQQGEwJDTjEOMAwGA1UECAwFU3RhdGUxCzAJBgNVBAcMAkNOMRUwEwYDVQQKDAxv
+cmdhbml6YXRpb24xHDAaBgNVBAsME29yZ2FuaXphdGlvbmFsIHVuaXQxFDASBgNV
+BAMMC2V4YW1wbGUuY29tMSIwIAYJKoZIhvcNAQkBFhNleGFtcGxlQGV4YW1wbGUu
+Y29tMB4XDTE2MDUxNjAyNDY1NVoXDTI2MDUxNDAyNDY1NVowgZkxCzAJBgNVBAYT
+AkNOMQ4wDAYDVQQIDAVTdGF0ZTELMAkGA1UEBwwCQ04xFTATBgNVBAoMDG9yZ2Fu
+aXphdGlvbjEcMBoGA1UECwwTb3JnYW5pemF0aW9uYWwgdW5pdDEUMBIGA1UEAwwL
+ZXhhbXBsZS5jb20xIjAgBgkqhkiG9w0BCQEWE2V4YW1wbGVAZXhhbXBsZS5jb20w
+ggIiMA0GCSqGSIb3DQEBAQUAA4ICDwAwggIKAoICAQC2ky/K/XneJKbCbpOsWlQ7
+OwgYEQNsa044RkwSbTwPwgLafUZ3r9c5nkXE8APqAikTQQBwyiNjk7QeXgIOjJXd
+7+IpwGoU6Bi2miA21qfvJPknyDAqw9tT/ycGQrvkY6rnqd++ri30ZUByUgO0du6+
+aWHo7af5/G1HQz0tu6i1tIF1dhSHNeqJKwxyUG8vIiT/PfbtU/mXSdQ07M+4ojBC
+O7FgoOS+rWgbL3yhWUTrCXSV2HZlhksYBhtWGoFVRPVSf89iqL02h9rZEjmfVY6R
+QlCnzu9v49Q8WFU528f+gDNXr9v13PKEDmloMzTqWPaCyD2FBbEKBsWHXHf1zqlI
+jyGZV7rHZ3i0C1LI6bdDDP7M7aVs8O+RjxK+HmfFRg5us2t6g7zAevwwLpMZRAud
+S39F91Up7l9g8WXpViok/8vcsOdePvvWcWro8qJhuEHAnDdMzj2Cko1L85/vRM/a
+budWXK7Ix0TlPWPfHJc2SLFeqqcm5Iypf/cGabQ6f0oRt6bCfspFgX9upznT5FwZ
+R0o1w6Q3q+4xVl6LgZvEAudWppyz79RACJA/jbXZQ7uJkXAxoI0nev9vgY6XJqUj
+XIQDih2hmi/uTnNU7Me7w7pCYKPdHlNU652kaJSH6W6ZFGk2rEOCOeAuWO9pZTq2
+3IhuOcDAKOcmimlkzaWRGQIDAQABo1AwTjAdBgNVHQ4EFgQUPJF++WMsv1OJvf7F
+oCew37JTnfQwHwYDVR0jBBgwFoAUPJF++WMsv1OJvf7FoCew37JTnfQwDAYDVR0T
+BAUwAwEB/zANBgkqhkiG9w0BAQsFAAOCAgEAb5LvqukMxWd5Zajbh3orfYsXmhWn
+UWiwG176+bd3b5xMlG9iLd4vQ11lTZoIhFOfprRQzbizQ8BzR2JBQckpLcy+5hyA
+D3M9vLL37OwA0wT6kxFnd6LtlFaH5gG++huw2ts2PDXFz0jqw+0YE/R8ov2+YdaZ
+aPSEMunmAuEY1TbYWzz4u6PxycxhQzDQ34ZmJZ34Elvw1NYMfPMGTKp34PsxIcgT
+ao5jqb9RMU6JAumfXrOvXRjjl573vX2hgMZzEU6OF2/+uyg95chn6nO1GUQrT2+F
+/1xIqfHfFCm8+jujSDgqfBtGI+2C7No+Dq8LEyEINZe6wSQ81+ryt5jy5SZmAsnj
+V4OsSIwlpR5fLUwrFStVoUWHEKl1DflkYki/cAC1TL0Om+ldJ219kcOnaXDNaq66
+3I75BvRY7/88MYLl4Fgt7sn05Mn3uNPrCrci8d0R1tlXIcwMdCowIHeZdWHX43f7
+NsVk/7VSOxJ343csgaQc+3WxEFK0tBxGO6GP+Xj0XmdVGLhalVBsEhPjnmx+Yyrn
+oMsTA1Yrs88C8ItQn7zuO/30eKNGTnby0gptHiS6sa/c3O083Mpi8y33GPVZDvBl
+l9PfSZT8LG7SvpjsdgdNZlyFvTY4vsB+Vd5Howh7gXYPVXdCs4k7HMyo7zvzliZS
+ekCw9NGLoNqQqnA=
+-----END CERTIFICATE-----
```
Where is that used? and is it good idea to keep it in the repository?

OrlinVasilev commented on config/core/private_key.pem:
```diff
@@ -0,0 +1,51 @@
+-----BEGIN RSA PRIVATE KEY-----
+MIIJKAIBAAKCAgEAtpMvyv153iSmwm6TrFpUOzsIGBEDbGtOOEZMEm08D8IC2n1G
+d6/XOZ5FxPAD6gIpE0EAcMojY5O0Hl4CDoyV3e/iKcBqFOgYtpogNtan7yT5J8gw
+KsPbU/8nBkK75GOq56nfvq4t9GVAclIDtHbuvmlh6O2n+fxtR0M9LbuotbSBdXYU
+hzXqiSsMclBvLyIk/z327VP5l0nUNOzPuKIwQjuxYKDkvq1oGy98oVlE6wl0ldh2
+ZYZLGAYbVhqBVUT1Un/PYqi9Nofa2RI5n1WOkUJQp87vb+PUPFhVOdvH/oAzV6/b
+9dzyhA5paDM06lj2gsg9hQWxCgbFh1x39c6pSI8hmVe6x2d4tAtSyOm3Qwz+zO2l
+bPDvkY8Svh5nxUYObrNreoO8wHr8MC6TGUQLnUt/RfdVKe5fYPFl6VYqJP/L3LDn
+Xj771nFq6PKiYbhBwJw3TM49gpKNS/Of70TP2m7nVlyuyMdE5T1j3xyXNkixXqqn
+JuSMqX/3Bmm0On9KEbemwn7KRYF/bqc50+RcGUdKNcOkN6vuMVZei4GbxALnVqac
+s+/UQAiQP4212UO7iZFwMaCNJ3r/b4GOlyalI1yEA4odoZov7k5zVOzHu8O6QmCj
+3R5TVOudpGiUh+lumRRpNqxDgjngLljvaWU6ttyIbjnAwCjnJoppZM2lkRkCAwEA
+AQKCAgAvsvCPlf2a3fR7Y6xNISRUfS22K+u7DaXX6fXB8qv4afWY45Xfex89vG35
+78L2Bi55C0h0LztjrpkmPeVHq88TtrJduhl88M5UFpxH93jUb9JwZErBQX4xyb2G
+UzUHjEqAT89W3+a9rR5TP74cDd59/MZJtp1mIF7keVqochi3sDsKVxkx4hIuWALe
+csk5hTApRyUWCBRzRCSe1yfF0wnMpA/JcP+SGXfTcmqbNNlelo/Q/kaga59+3UmT
+C0Wy41s8fIvP+MnGT2QLxkkrqYyfwrWTweqoTtuKEIHjpdnwUcoYJKfQ6jKp8aH0
+STyP5UIyFOKNuFjyh6ZfoPbuT1nGW+YKlUnK4hQ9N/GE0oMoecTaHTbqM+psQvbj
+6+CG/1ukA5ZTQyogNyuOApArFBQ+RRmVudPKA3JYygIhwctuB2oItsVEOEZMELCn
+g2aVFAVXGfGRDXvpa8oxs3Pc6RJEp/3tON6+w7cMCx0lwN/Jk2Ie6RgTzUycT3k6
+MoTQJRoO6/ZHcx3hTut/CfnrWiltyAUZOsefLuLg+Pwf9GHhOycLRI6gHfgSwdIV
+S77UbbELWdscVr1EoPIasUm1uYWBBcFRTturRW+GHJ8TZX+mcWSBcWwBhp15LjEl
+tJf+9U6lWMOSB2LvT+vFmR0M9q56fo7UeKFIR7mo7/GpiVu5AQKCAQEA6Qs7G9mw
+N/JZOSeQO6xIQakC+sKApPyXO58fa7WQzri+l2UrLNp0DEQfZCujqDgwys6OOzR/
+xg8ZKQWVoad08Ind3ZwoJgnLn6QLENOcE6PpWxA/JjnVGP4JrXCYR98cP0sf9jEI
+xkR1qT50GbeqU3RDFliI4kGRvbZ8cekzuWppfQcjstSBPdvuxqAcUVmTnTw83nvD
+FmBbhlLiEgI3iKtJ97UB7480ivnWnOuusduk7FO4jF3hkrOa+YRidinTCi8JBo0Y
+jx4Ci3Y5x6nvwkXhKzXapd7YmPNisUc5xA7/a+W71cyC0IKUwRc/8pYWLL3R3CpR
+YiV8gf6gwzOckQKCAQEAyI9CSNoAQH4zpS8B9PF8zILqEEuun8m1f5JB3hQnfWzm
```
where is that cert used? and why we are having it in the repository?

OrlinVasilev commented on config/core/private_key.pem:
```diff
@@ -0,0 +1,51 @@
+3R5TVOudpGiUh+lumRRpNqxDgjngLljvaWU6ttyIbjnAwCjnJoppZM2lkRkCAwEA
+AQKCAgAvsvCPlf2a3fR7Y6xNISRUfS22K+u7DaXX6fXB8qv4afWY45Xfex89vG35
+78L2Bi55C0h0LztjrpkmPeVHq88TtrJduhl88M5UFpxH93jUb9JwZErBQX4xyb2G
+UzUHjEqAT89W3+a9rR5TP74cDd59/MZJtp1mIF7keVqochi3sDsKVxkx4hIuWALe
+csk5hTApRyUWCBRzRCSe1yfF0wnMpA/JcP+SGXfTcmqbNNlelo/Q/kaga59+3UmT
+C0Wy41s8fIvP+MnGT2QLxkkrqYyfwrWTweqoTtuKEIHjpdnwUcoYJKfQ6jKp8aH0
+STyP5UIyFOKNuFjyh6ZfoPbuT1nGW+YKlUnK4hQ9N/GE0oMoecTaHTbqM+psQvbj
+6+CG/1ukA5ZTQyogNyuOApArFBQ+RRmVudPKA3JYygIhwctuB2oItsVEOEZMELCn
+g2aVFAVXGfGRDXvpa8oxs3Pc6RJEp/3tON6+w7cMCx0lwN/Jk2Ie6RgTzUycT3k6
+MoTQJRoO6/ZHcx3hTut/CfnrWiltyAUZOsefLuLg+Pwf9GHhOycLRI6gHfgSwdIV
+S77UbbELWdscVr1EoPIasUm1uYWBBcFRTturRW+GHJ8TZX+mcWSBcWwBhp15LjEl
+tJf+9U6lWMOSB2LvT+vFmR0M9q56fo7UeKFIR7mo7/GpiVu5AQKCAQEA6Qs7G9mw
+N/JZOSeQO6xIQakC+sKApPyXO58fa7WQzri+l2UrLNp0DEQfZCujqDgwys6OOzR/
+xg8ZKQWVoad08Ind3ZwoJgnLn6QLENOcE6PpWxA/JjnVGP4JrXCYR98cP0sf9jEI
+xkR1qT50GbeqU3RDFliI4kGRvbZ8cekzuWppfQcjstSBPdvuxqAcUVmTnTw83nvD
+FmBbhlLiEgI3iKtJ97UB7480ivnWnOuusduk7FO4jF3hkrOa+YRidinTCi8JBo0Y
+jx4Ci3Y5x6nvwkXhKzXapd7YmPNisUc5xA7/a+W71cyC0IKUwRc/8pYWLL3R3CpR
+YiV8gf6gwzOckQKCAQEAyI9CSNoAQH4zpS8B9PF8zILqEEuun8m1f5JB3hQnfWzm
+7uz/zg6I0TkcCE0AJVSKPHQm1V9+TRbF9+DiOWHEYYzPmK8h63SIufaWxZPqai4E
+PUj6eQWykBUVJ96n6/AW0JHRZ+WrJ5RXBqCLuY7NP6wDhORrCJjBwaGMohNpbKPS
+H3QewsoxCh+CEXKdKyy+/yU/f4E89PlHapkW1/bDJ5u7puSD+KvmiDDIXSBncdOO
+uFT8n+XH5IwgjdXFSDim15rQ8jD2l2xLcwKboTpx5GeRl8oB1VGm0fUbBn1dvGPG
+4WfHGyrp9VNZtP160WoHr+vRVPqvHNkoeAlCfEwQCQKCAQBN1dtzLN0HgqE8TrOE
+ysEDdTCykj4nXNoiJr522hi4gsndhQPLolb6NdKKQW0S5Vmekyi8K4e1nhtYMS5N
+5MFRCasZtmtOcR0af87WWucZRDjPmniNCunaxBZ1YFLsRl+H4E6Xir8UgY8O7PYY
+FNkFsKIrl3x4nU/RHl8oKKyG9Dyxbq4Er6dPAuMYYiezIAkGjjUCVjHNindnQM2T
+GDx2IEe/PSydV6ZD+LguhyU88FCAQmI0N7L8rZJIXmgIcWW0VAterceTHYHaFK2t
+u1uB9pcDOKSDnA+Z3kiLT2/CxQOYhQ2clgbnH4YRi/Nm0awsW2X5dATklAKm5GXL
+bLSRAoIBAQClaNnPQdTBXBR2IN3pSZ2XAkXPKMwdxvtk+phOc6raHA4eceLL7FrU
+y9gd1HvRTfcwws8gXcDKDYU62gNaNhMELWEt2QsNqS/2x7Qzwbms1sTyUpUZaSSL
+BohLOKyfv4ThgdIGcXoGi6Z2tcRnRqpq4BCK8uR/05TBgN5+8amaS0ZKYLfaCW4G
+nlPk1fVgHWhtAChtnYZLuKg494fKmB7+NMfAbmmVlxjrq+gkPkxyqXvk9Vrg+V8y
+VIuozu0Fkouv+GRpyw4ldtCHS1hV0eEK8ow2dwmqCMygDxm58X10mYn2b2PcOTl5
+9sNerUw1GNC8O66K+rGgBk4FKgXmg8kZAoIBABBcuisK250fXAfjAWXGqIMs2+Di
+vqAdT041SNZEOJSGNFsLJbhd/3TtCLf29PN/YXtnvBmC37rqryTsqjSbx/YT2Jbr
+Bk3jOr9JVbmcoSubXl8d/uzf7IGs91qaCgBwPZHgeH+kK13FCLexz+U9zYMZ78fF
+/yO82CpoekT+rcl1jzYn43b6gIklHABQU1uCD6MMyMhJ9Op2WmbDk3X+py359jMc
++Cr2zfzdHAIVff2dOV3OL+ZHEWbwtnn3htKUdOmjoTJrciFx0xNZJS5Q7QYHMONj
+yPqbajyhopiN01aBQpCSGF1F1uRpWeIjTrAZPbrwLl9YSYXz0AT05QeFEFk=
+-----END RSA PRIVATE KEY-----
```
where is this one used? and why we are keeping private key in the repository?!?!

ShrutiC-git commented:
hi @OrlinVasilev ~ thanks for the review on this. Agree, the certs shouldn't be needed. I had added this for the config of the services, but these can be removed. I'll be cleaning the PR according to the suggestions (and will also point to some necessary changes, added to core code)

ShrutiC-git commented:
hi, @Vad1mo @OrlinVasilev, I have moved preview-specific files to their own directory to keep the integration code separate. Also, as you pointed out, have removed the files that are not used in the service volumes. Thanks for helping clean this up!

Let me know how I can make it more efficient. We can probably add more modules to be built from source, but some of the co-dependencies were missing so currently we are limited to building 3 of Harbor components.

OrlinVasilev commented:
@ShrutiC-git how we can test it ?

ShrutiC-git commented:
hey, @OrlinVasilev -

To test this, here's what I had done: 

The first step was opening a PR against my fork of Harbor, the uffizzi-build.yml file triggered the Build PR Image action. Once this action was completed, the uffizzi-preview.yml GHA-file triggered the Deploy Preview action. We use this 2-step workflow for open-source projects, to allow open-source projects to have previews for contributions from forks. [Here](https://www.uffizzi.com/preview-environments-guide/github-actions-two-stage-workflow) is a more in-depth read into why we use a two-stage workflow.

* [Here is the PR](https://github.com/ShrutiC-git/harbor/pull/8) opened against my personal fork of Habor
* [Here is the comment](https://github.com/ShrutiC-git/harbor/pull/8#issuecomment-1398012998) that was posted on my PR.
* [Here is the preview environment](https://app.uffizzi.com/github.com/ShrutiC-git/harbor/pull/8) that was created.

To log into the preview, use the following credentials:
* username: admin
* password: password


**For previews on this repo, this PR needs to be merged. That is because the `Deploy Preview action` is a `workflow_run` type of action and needs the action file (`uffizzi-preview.yml`) to be in the `default` branch of the PR.** 

[Here's an in-depth read](https://www.uffizzi.com/preview-environments-guide/github-actions-two-stage-workflow) on why we use the 2-stage workflow, if you'd like to know why we have the workflows separated out into two separate actions.

OrlinVasilev commented:
@ShrutiC-git - the backstage environment looks quite good and responsive, is there a way to make that available for tests with Harbor and possibly to do some demo on our next community meeting thanks!

ShrutiC-git commented:
hey @OrlinVasilev, thanks for looking further into this.

> is there a way to make that available for tests with Harbor

could you maybe explain what you mean by this? What would be helpful for the community? 

were you able to take a look at the preview environment link for Harbor (linked in the comment above)? How does that look so far?

OrlinVasilev commented:
@ShrutiC-git thank you!

@goharbor/all-maintainers can you take a look here: Thank you!
https://app.uffizzi.com/github.com/ShrutiC-git/harbor/pull/8

username: admin
password: password

ShrutiC-git commented:
monitoring this too, @OrlinVasilev. 

Let me know what changes we can make to these ephemeral environments to align with what the community needs.

OrlinVasilev commented:
@jpthurman - waiting the rest of the @goharbor/all-maintainers to test it and share some thought
It's great for testing UI changes, for the core components not so much as we cannot access the underneath containers/pods and check logs and troubleshoot if needed
But I like it for the UI/UX side it's solid

BTW do you plan to leave it free for the OS projects or we can expect some changes in the policy in the future?
Orlix

OrlinVasilev commented:
as a follow up question can we make it on demand so we can exclude the possibility our pipelines to depend on it?

jpthurman commented:
Hi @OrlinVasilev 
-You can access each container's logs via the Uffizzi UI - are you needing deeper access at the infrastructure level?
-If this is purely open source and not tied to a profit mechanism we can keep it free.  If you could provide special recognition in your README that would be quite helpful
-Can you elaborate on what you would like to see with "on-demand"?

jpthurman commented:
Hi @OrlinVasilev 

Great - yes the service will be free for Harbor

For on-demand yes you can set the GHA for whatever scenarios you can imagine.  What's in this PR is the template used across many Open Source projects and they are quite happy with it.  

How much longer do you plan to get feedback from harbor community?  Do you want to set a time period for feedback?

When you are ready we can update this PR (I'm sure it's a little stale) and you can merge.  Tagging @ShrutiC-git

OrlinVasilev commented:
@ShrutiC-git @jpthurman hi folks :) are you still interested to do a live session during our community meetings :) so we can finally decide and merge this one :))

jpthurman commented:
Hi @OrlinVasilev!
Given where we are timing-wise I think the project would benefit longer term from our pending helm support.  Would it work for us to update the PR in the next 60 days or so and then demo?

OrlinVasilev commented:
@jpthurman fantastic! Let's park that for now then! BTW do you need any help with the helm part maybe I can contribute :)

OrlinVasilev commented:
@jpthurman will be waiting :)
